### PR TITLE
[tado] assorted upgrades and bug fixes to the binding

### PR DESCRIPTION
**Pull request by andrewfg**
This PR provides assorted bug fixes and enhancements to the tado binding as follows..

- Replaced the List of Zones by a HashMap of Zone Ids vs. Battery state, to avoid making a JSON HTTP request to the API every time that Battery state is refreshed; so now the battery state only gets polled after the polling refresh time has expired.

- For Battery state and Open Window alarms, the binding only returns a JSON payload element in the case that there IS actually an alarm, (i.e. if there is no JSON element this means there is no alarm); so in such cases the Channels now return OnOffType.OFF (no alarm) instead of UndefType.UNDEF.

- There were typo errors in the POM.XML dependencies entry: a) artifactId should be "tado-api-client" (not "api-client"), and b) version should be "1.3" (not "1.3.0") -- NOTE: I think this error was introduced in [#6183](https://github.com/openhab/openhab-addons/pull/6183)

- Added pom-aggregator.xml to the root directory in order to build the "tado-api-client-1.3.jar" dependency file from the root => hopefully the reviewers can suggest a way to integrate this build directly into the OH MVN build environment. (??)

### CODE REVIEW

**Comment by andrewfg** [comment]
Note: the Travis tests have failed. This is due to the aforementioned error in the POM.XML file (which I have corrected). However the correct dependency jar file "tado-api-client-1.3.jar" needs to be added to the OH MVN repository.

> There were typo errors in the POM.XML dependencies entry: a) artifactId should be "tado-api-client" (not "api-client"), and b) version should be "1.3" (not "1.3.0") -- NOTE: I think this error was introduced in #6183

**Comment by J-N-K** [comment]
I don't think this is the problem. The file is available and the naming shouldn't be a problem.

https://bintray.com/openhab/mvn/openhab-addons-deps#files/org%2Fopenhab%2Fbinding%2Ftado%2Fapi-client%2F1.3.0

Why do you think that it makes a difference?

File: bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java [hunk 1]
```diff
@@ -57,10 +57,20 @@ private synchronized void refreshZoneList() {
             Long homeId = homeHandler.getHomeId();
             if (homeId != null) {
                 logger.debug("Fetching (battery state) zone list for HomeId {}", homeId);
+                zoneList.clear();
                 try {
-                    zoneList = homeHandler.getApi().listZones(homeId);
+                    for (Zone zone : homeHandler.getApi().listZones(homeId)) {
```
Comment (J-N-K) [inline bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java]: What's the idea here? If any `batteryState` is low, set the `zoneList` to `OFF` otherwise to `ON`?

```
homeHandler.getApi().listZones(homeId).forEach(zone -> {
    try {
        boolean batteryLow = !zone.getDevices().stream().map(ControlDevice::getBatteryState).anyMatch(s->!("NORMAL".equals(s)));
        zoneList.put(Long:valueOf(zone.getId()), OnOffType.from(batteryLow));
    } catch (IOException | ApiException e) {
        logger.debug("Fetch (battery state) zone list exception", e); 
    }
}
```

This also allows to process other zones if one zone failed (instead of before where processing stops when the first error occurs.

File: bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java [hunk 2]
```diff
@@ -69,25 +79,8 @@ private synchronized void refreshZoneList() {
...
-                            }
-                            hasBatteryStateValue = true;
-                        }
-                    }
-                    break;
-                }
-            }
-        }
-        return hasBatteryStateValue ? OnOffType.OFF : UnDefType.UNDEF;
+        State state = zoneList.get(zoneId);
```
Comment (J-N-K) [inline bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java]:
```suggestion
        return zoneList.getorDefault(zoneId, OnOffType.OFF);
```

**Comment by andrewfg** [comment]
> I don't think this is the problem. The file is available and the naming shouldn't be a problem.
> Why do you think that it makes a difference?

@J-N-K, it makes a difference simply because we must have consistent names for the jar. Currently the same jar is called by three different names as follows..

1. In src/main/api there is a pom.xml that builds the jar, which has artifactId="tado-client-api" and version="1.3" so it builds a jar named "**_tado-client-api-1.3.jar_**" (with "tado-" prefix / no ".0" suffix)

2. In the binding root pom.xml it specifies a dependency on artifactId="client-api" and version="1.3.0" i.e. "**_client-api-1.3.0.jar_**" (no "tado-" prefix / with ".0" suffix)

3. In the openHAB dependency files store there is a file called "**_tado-client-api-1.3.0.jar_**" (with "tado-" prefix / with ".0" suffix)

IMHO we must resolve this confusion.

**Comment by J-N-K** [comment]
Where does (3) come from? I don't have that. Maybe it's a relict from pre-#6183?

File: bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java (hunk 1 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java]: I think I need to give you some background here..

In a tado system the Bridge Thing is the home "HomeId", and within each Home, there are Things for rooms "ZoneId" (for example Living Room Heating). In the majority of cases, each Zone has a single radiator control valve device in it. But the tado system does also allow for each Zone to have more than one radiator control valve; and in that case the two (or more) radiator control valves are slaved together under the single Zone Thing.
```
HomeId = "my address"
├── ZoneId = "Kitchen"
│   ├── DeviceID = "radiator control valve A"
├── ZoneId = "Living Room"
     ├── DeviceID = "radiator control valve B"
     ├── DeviceID = "radiator control valve C"
```
In the binding, the main API interface is a JSON request to http://something/something/ZoneId=J which fetches Channel information for the Zone like room temperature, set temperature, heating state. The binding polls this JSON request according to the refresh Interval defined in the Configuration Parameters.

HOWEVER, for the low battery alarm -- unlike the other Channels -- the checking for this alarm is NOT done at Zone level but rather at each radiator control valve device level. So for each device there is a separate JSON request to the API as http://something/something/DeviceId=K which reads the low battery alarm state.

The radiator control valve devices are not modelled as Things in the binding. So to bring their low battery alarm state into openHAB it is necessary to poll each device, and amalgamate these alarm values in the common Zone low battery battery alarm Channel (by OR'ing across all the radiator control valve devices in that Zone).

Furthermore, in the previous build of this binding, the JSON calls for the low battery alarm were not being made according to the polling refresh Interval, but were instead being made immediately every time that the binding requested an update. So this caused far too many unnecessary API calls -- particularly since this Channel only goes into alarm roughly once in 6 months.

So in this new build of the binding, I have done two things: a) the polling of JSON request to the radiator control valve devices for their low battery alarm state is no longer made on every refresh request, and b) there is now a Map table that maps between the Zone Ids and the amalgamated low battery alarm states of the radiator control valve devices within that Zone.

**Comment by andrewfg** [comment]
> Where does (3) come from? I don't have that. 

Err... I'm just quoting what you said earlier in this thread..

> The file is available and the naming shouldn't be a problem.
>https://bintray.com/openhab/mvn/openhab-addons-deps#files/org%2Fopenhab%2Fbinding%2Ftado%2Fapi-client%2F1.3.0

**Comment by J-N-K** [comment]
The file I referenced above is exactly the file that is in the current pom, and has exactly the name it has remote in my local maven cache (i.e. api-client-1.3.0.jar). If we compile the file ourselves, why don't we follow the semver-scheme (i.e. major.minor.micro)?

File: bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java (hunk 1 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java]: I think that is exactly what my code-snippet is doing. It's only another way of doing the same thing.

**Comment by andrewfg** [comment]
> The file I referenced above is exactly the file that is in the current pom, and has exactly the name it has remote in my local maven cache (i.e. api-client-1.3.0.jar).

You are right. Sorry for that. So we only have two different names to resolve rather than three :)

> If we compile the file ourselves, why don't we follow the semver-scheme (i.e. major.minor.micro)? I would recommend changing the api/pom.xml to 1.3.0.

I have no problem with that. In that case I would revert the proposed renaming of the dependency in the root pom.xml (the pom that builds the binding) and instead I would change the artifactId and version in /src/main/api/pom.xml (the pom that builds the lib jar) to match => Is that Ok with you?

An additional question is if there is some way to modify the pom.xml in the root directory so it also calls /src/main/api/pom.xml so that the OH MVN build reactor can build the lib jar at the same time as it builds the binding jar? Or combine the actions of the two pom's into one file in the root folder? (Unfortunately I am no great expert in the syntax of pom files ..)

**Comment by andrewfg** [comment]
PS and I think that either 

a) if the lib jar should be built on your side, then .gitignore must be modified to exclude from future PRs any lib jar built by the contributor, OR..

b) If the lib jar will be built by the contributor then there needs to be an automated way to add the lib jar into the dependencies store your side..

**Comment by J-N-K** [comment]
1. Yes, please change the POM to match what we currently have.

2. In general it should be possible to integrate that. However, it is a bit tricky to get it right. We must make sure that the library

- is build before the rest of the binding
- is available on the classpath so it can be resolved from the IDE
- is not committed as we don't allow binary files in the repo any longer
- is included in the jar

I would need a closer look on how to do that, unfortunately my time is quite limited ATM. I would suggest to keep that in mind for a later improvement.

File: bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java (hunk 2 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java]: or probably default to `UndefType.UNDEF` as we don't know if `ON` or `OFF` is correct.

**Comment by andrewfg** [comment]
```
homeHandler.getApi().listZones(homeId).forEach(zone -> {
    try {
        boolean batteryLow = !zone.getDevices().stream().map(ControlDevice::getBatteryState).anyMatch(s->!("NORMAL".equals(s)));
        zoneList.put(Long:valueOf(zone.getId()), OnOffType.from(batteryLow));
    } catch (IOException | ApiException e) {
        logger.debug("Fetch (battery state) zone list exception", e); 
    }
}
```
>This also allows to process other zones if one zone failed (instead of before where processing stops when the first error occurs.

@J-N-K no, it is the listZones(homeId) method that might throw the exceptions, so the try/catch has to surround listZones(homeId); and (therefore) a catch for exceptions inside the for/foreach code does not compile resp. does not work.

PS I really like your cool one line code (below) -- but I I need to think about it a bit first (??)

```
boolean batteryLow = !zone.getDevices().stream().map(ControlDevice::getBatteryState).anyMatch(s->!("NORMAL".equals(s)));
```

**Comment by andrewfg** [comment]
```
boolean batteryLow = !zone.getDevices().stream().map(ControlDevice::getBatteryState).anyMatch(s->!("NORMAL".equals(s)));
```
I think this code does not actually deliver the logic result that is needed; the needed logic outcomes are as follows..

- JSON element for batteryState NOT present (i.e. batteryState == null) => No alarm
- JSON element for batteryState == "NORMAL" => No alarm
- JSON element for batteryState != "NORMAL" => Alarm !!

File: bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java (hunk 2 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java]: Agreed to your two points. I was not aware of the getOrDefault() method, so I definitely learned something useful, thank you.

**Comment by J-N-K** [comment]
Maybe this one.

```
try {
    homeHandler.getApi().listZones(homeId).forEach(zone -> {
        boolean batteryLow = !zone.getDevices().stream().map(ControlDevice::getBatteryState).filter(Objects::nonNull).allMatch(s->s.equals("NORMAL"));
        zoneList.put(Long:valueOf(zone.getId()), OnOffType.from(batteryLow));
    });
} catch (IOException | ApiException e) {
    logger.debug("Fetch (battery state) zone list exception", e); 
}
```

**Comment by andrewfg** [comment]
```
boolean batteryLow = !zone.getDevices().stream().map(ControlDevice::getBatteryState)
  .filter(Objects::nonNull).allMatch(s->s.equals("NORMAL"));
```
Yup! That works. Many Thanks.

I think the PR is ready to go now => so long as you are happy?

File: bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java (hunk 1 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java]: Resolved. See comments below..

File: bundles/org.openhab.binding.tado/pom-aggregator.xml [hunk 3]
```diff
@@ -0,0 +1,16 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?><project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
```
Comment (J-N-K) [inline bundles/org.openhab.binding.tado/pom-aggregator.xml]: Is this file still needed? IMO we should try to integrate it in the main pom (in another PR)

File: bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java [hunk 4]
```diff
@@ -57,37 +58,23 @@ private synchronized void refreshZoneList() {
...
-                    zoneList = homeHandler.getApi().listZones(homeId);
+                    homeHandler.getApi().listZones(homeId).forEach(zone -> {
+                        boolean batteryLow = !zone.getDevices().stream().map(ControlDevice::getBatteryState)
+                                .filter(Objects::nonNull).allMatch(s -> s.equals("NORMAL"));
+                        zoneList.put(Long.valueOf(zone.getId()), OnOffType.from(batteryLow));
+                    });
                 } catch (IOException | ApiException e) {
-                    zoneList = null;
-                    logger.debug("Fetch (battery state) zone list exception", e);
+                    logger.debug("Fetch (battery state) zone list exception");
```
Comment (J-N-K) [inline bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java]: Did you remove the stacktrace by intention?

File: bundles/org.openhab.binding.tado/pom-aggregator.xml (hunk 3 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.tado/pom-aggregator.xml]: Yes it’s ok to delete it. Shall l do it or will you?

File: bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java (hunk 4 above)
Comment (andrewfg) [reply bundles/org.openhab.binding.tado/src/main/java/org/openhab/binding/tado/internal/handler/TadoBatteryChecker.java]: Yes. :)

File: bundles/org.openhab.binding.tado/pom-aggregator.xml (hunk 3 above)
Comment (J-N-K) [reply bundles/org.openhab.binding.tado/pom-aggregator.xml]: would be great if you could do that

Comment (andrewfg) [reply bundles/org.openhab.binding.tado/pom-aggregator.xml]: Done
